In [54]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

% matplotlib inline

In [5]:
ds = pd.read_csv('./datasets/properati_argentina_2021.csv')
houses = ds.copy()
houses.head()

id  start_date    end_date  created_on    latitud  \
0  XZq2dFGxBiboHY3Bo/tqlw==  2021-09-09  2022-01-27  2021-09-09 -38.957632   
1  8afoK+yMH/rnnlYiktwgNA==  2021-09-09  2021-10-23  2021-09-09 -34.888878   
2  iVaG5rBB47dd1Ya4BMXREA==  2021-09-09  2022-03-09  2021-09-09 -32.488717   
3  QE1QVMWKAHWkd0qT0FtPWA==  2021-09-09  2022-01-27  2021-09-09 -34.900165   
4  tJ5lSk7IysG/71OGhacNZA==  2021-09-09  2022-01-27  2021-09-09 -38.954528   

    longitud    place_l2 place_l3 place_l4 place_l5  place_l6 operation  \
0 -68.105443     Neuquén      NaN      NaN      NaN       NaN     Venta   
1 -56.092368  Montevideo      NaN      NaN      NaN       NaN     Venta   
2 -58.257499  Entre Ríos      NaN      NaN      NaN       NaN     Venta   
3 -54.808453   Maldonado      NaN      NaN      NaN       NaN     Venta   
4 -68.041862     Neuquén      NaN      NaN      NaN       NaN     Venta   

  property_type  property_rooms  property_bedrooms  property_surface_total  \
0          Casa             6.0                NaN                     NaN   
1          Casa             4.0                3.0                   193.0   
2          Casa             NaN                NaN                   110.0   
3          Casa             NaN                3.0                     NaN   
4            PH             3.0                2.0                     NaN   

   property_surface_covered  property_price property_currency  \
0                     206.0             NaN               NaN   
1                     193.0        390000.0               USD   
2                      45.0         47000.0               USD   
3                     180.0        400000.0               USD   
4                      98.0        102000.0               USD   

       property_title  
0      Casa - Capital  
1       Casa - Malvín  
2      Casa - Uruguay  
3    Casa - El Chorro  
4  PH - Villa Farrell

In [6]:
houses.shape

(460154, 20)

In [7]:
houses.describe()

latitud       longitud  place_l6  property_rooms  \
count  419740.000000  419740.000000       0.0   368498.000000   
mean      -34.228570     -59.557066       NaN        3.278053   
std         3.251509       2.905129       NaN        1.870832   
min       -54.832057    -103.219258       NaN        1.000000   
25%       -34.640132     -58.899343       NaN        2.000000   
50%       -34.585122     -58.488309       NaN        3.000000   
75%       -34.423414     -58.404876       NaN        4.000000   
max        42.557773     -35.044991       NaN       40.000000   

       property_bedrooms  property_surface_total  property_surface_covered  \
count      344113.000000           397813.000000              4.279160e+05   
mean            2.297960              420.601250              9.605005e+03   
std             1.673985             4026.380748              3.440367e+06   
min            -3.000000               -1.000000             -3.000000e+00   
25%             1.000000               50.000000              4.500000e+01   
50%             2.000000               90.000000              7.800000e+01   
75%             3.000000              210.000000              1.700000e+02   
max           390.000000           200000.000000              2.147484e+09   

       property_price  
count    4.421530e+05  
mean     3.063273e+05  
std      4.899613e+06  
min      0.000000e+00  
25%      4.490000e+04  
50%      9.800000e+04  
75%      1.990000e+05  
max      1.500000e+09

In [21]:
houses.columns

Index(['id', 'start_date', 'end_date', 'created_on', 'latitud', 'longitud',
       'place_l2', 'place_l3', 'place_l4', 'place_l5', 'place_l6', 'operation',
       'property_type', 'property_rooms', 'property_bedrooms',
       'property_surface_total', 'property_surface_covered', 'property_price',
       'property_currency', 'property_title'],
      dtype='object')

In [8]:
filas_totales = houses.shape[0]
print(houses.isna().sum() / filas_totales * 100)

id                            0.000000
start_date                    0.000000
end_date                      0.000000
created_on                    0.000000
latitud                       8.782712
longitud                      8.782712
place_l2                      0.000000
place_l3                      4.887277
place_l4                     69.788375
place_l5                     99.471916
place_l6                    100.000000
operation                     0.000000
property_type                 0.000000
property_rooms               19.918549
property_bedrooms            25.217862
property_surface_total       13.547856
property_surface_covered      7.005915
property_price                3.911951
property_currency             4.034302
property_title                0.000000
dtype: float64


## Análisis de la variable precio

In [13]:
houses[houses.property_price.isna()].shape
# 18 mil casas sin precio

(18001, 20)

Hay 18 mil casas sin precio, representan un 3% de los datos, como queremos precedir el precio estas filas no nos sirven.

In [14]:
houses.dropna(subset=['property_price'], inplace=True)

In [15]:
# Verifico que haya bajado el porcentaje
filas_totales = houses.shape[0]
print(houses.isna().sum() / filas_totales * 100)

id                            0.000000
start_date                    0.000000
end_date                      0.000000
created_on                    0.000000
latitud                       8.836986
longitud                      8.836986
place_l2                      0.000000
place_l3                      4.762379
place_l4                     69.351333
place_l5                     99.462629
place_l6                    100.000000
operation                     0.000000
property_type                 0.000000
property_rooms               19.258040
property_bedrooms            24.853840
property_surface_total       13.405088
property_surface_covered      6.936513
property_price                0.000000
property_currency             0.127331
property_title                0.000000
dtype: float64


In [17]:
houses[houses.property_price == 0].shape

(563, 20)

Además hay 563 filas con el precio en 0, vamos a borrarlas por el mismo motivo. También convierto a numérico el precio por las dudas

In [28]:
houses.property_price = pd.to_numeric(houses.property_price)
houses.drop(houses[houses.property_price == 0].index, inplace=True)
houses[houses.property_price.isna()].shape, houses[houses.property_price == 0].shape

((0, 18), (0, 18))

In [36]:
houses.groupby('property_currency')['id'].count() / houses['id'].count()

property_currency
ARS    0.201248
COP    0.000027
PEN    0.000224
USD    0.790561
UYU    0.007939
Name: id, dtype: float64

Las múltiples monedas podrían alterar la escala de los gráficos. Vamos a agrupar y trabajar momentaneamente sobre el grupo en dolares.

In [39]:
houses[['property_currency', 'property_price']].groupby('property_currency').describe()

property_price                                         \
                           count          mean           std        min   
property_currency                                                         
ARS                      88869.0  6.071652e+05  6.561089e+06     4400.0   
COP                         12.0  5.914292e+08  4.678378e+08  1150000.0   
PEN                         99.0  1.725796e+06  4.599244e+06      750.0   
USD                     349104.0  2.109385e+05  5.061346e+05       50.0   
UYU                       3506.0  1.648302e+05  4.926249e+06     3500.0   

                                                                        
                           25%          50%          75%           max  
property_currency                                                       
ARS                    25000.0      40000.0      80000.0  1.111111e+09  
COP                297500000.0  422500000.0  785000000.0  1.500000e+09  
PEN                   159000.0     389900.0    1242000.0  3.175200e+07  
USD                    64000.0     120000.0     225000.0  5.000000e+07  
UYU                    18000.0      22000.0      26000.0  2.775000e+08

### Analisis de Lugar

lat - Latitud.
lon - Longitud.
l1 - Nivel administrativo 1: país.
l2 - Nivel administrativo 2: usualmente provincia.
l3 - Nivel administrativo 3: usualmente ciudad.
l4 - Nivel administrativo 4: usualmente barrio.

In [22]:
# Cantidad de datos nulos en los datos de ubicacion
houses[['place_l2', 'place_l3', 'place_l4', 'place_l5', 'place_l6']].isna().sum() / houses.shape[0]

place_l2    0.000000
place_l3    0.047648
place_l4    0.693474
place_l5    0.994624
place_l6    1.000000
dtype: float64

Podemos descartar place_l6 y place_l5 porque tienen practicamente todos los datos nulos

In [23]:
houses.drop(columns=['place_l5', 'place_l6'], inplace=True)

In [24]:
houses.place_l4.value_counts()

La Plata                    9676
Nordelta                    8464
Olivos                      3925
Ramos Mejía                 3351
Palermo Hollywood           2935
                            ... 
Bosque Real Country Club       1
José C Paz                     1
Comodoro Py                    1
Cachoeira do Bom Jesus         1
Juan José Paso                 1
Name: place_l4, Length: 785, dtype: int64

In [25]:
houses.place_l3.value_counts().head(10)

Rosario          32185
Palermo          25406
Tigre            23853
La Plata         19164
Pilar            15432
Belgrano         14150
Córdoba          12628
Recoleta         11624
Escobar          11269
Vicente López    11074
Name: place_l3, dtype: int64

In [26]:
houses.place_l2.value_counts()

Capital Federal                 154788
Bs.As. G.B.A. Zona Norte         88213
Bs.As. G.B.A. Zona Sur           46168
Santa Fe                         42877
Bs.As. G.B.A. Zona Oeste         27852
Córdoba                          21676
Buenos Aires Costa Atlántica     11338
Neuquén                           7536
Montevideo                        7436
Buenos Aires Interior             5794
Maldonado                         5235
Río Negro                         4545
Mendoza                           4457
Tucumán                           3229
Colonia                           2433
Salta                             1416
Entre Ríos                        1045
San Juan                           845
Canelones                          738
San Luis                           575
Chubut                             560
La Pampa                           558
Misiones                           512
Florida                            405
Jujuy                              374
Corrientes               

In [95]:
houses[['place_l2', 'place_l3', 'property_price']][houses.place_l2 == 'Capital Federal']



place_l2          154788
place_l3          153817
property_price    154788
dtype: int64

In [71]:
grouped = houses[['property_currency', 'property_type', 'property_price']].groupby('property_currency')


In [70]:


# create color mapping based on all unique values of cat
unique_cat = grouped.  # get unique cats
colors = sns.color_palette('tab10', n_colors=len(unique_cat))  # get a number of colors
cmap = dict(zip(unique_cat, colors))  # zip values to colors

col_nums = 3  # how many plots per row
row_nums = math.ceil(len(df_dict) / col_nums)  # how many rows of plots

# create the figue and axes
fig, axes = plt.subplots(row_nums, col_nums, figsize=(9, 6), sharex=True, sharey=True)
# convert to 1D array for easy iteration
axes = axes.flat

# iterate through dictionary and plot
for ax, (k, v) in zip(axes, df_dict.items()):
    sns.scatterplot(data=v, x='x', y='y', hue='cat', palette=cmap, ax=ax)
    sns.despine(top=True, right=True)
    ax.legend_.remove()  # remove the individual plot legends
    ax.set_title(f'dataset = {k}', fontsize=11)

fig.tight_layout()

# create legend from cmap
# patches = [Patch(color=v, label=k) for k, v in cmap.items()]  # square patches
patches = [Line2D([0], [0], marker='o', color='w', markerfacecolor=v, label=k, markersize=8) for k, v in
           cmap.items()]  # round markers

# place legend outside of plot; change the right bbox value to move the legend up or down
plt.legend(title='cat', handles=patches, bbox_to_anchor=(1.06, 1.2), loc='center left', borderaxespad=0, frameon=False)
plt.show()
#    g.plot.box(title='Precio por tipo de propiedad', column='property_price', by='property_type')

{'ARS': [15, 18, 20, 22, 23, 24, 30, 40, 45, 46, 47, 48, 50, 52, 54, 72, 73, 82, 84, 85, 118, 127, 133, 153, 168, 169, 214, 219, 224, 230, 237, 257, 263, 271, 280, 284, 287, 290, 296, 300, 302, 307, 310, 311, 313, 314, 321, 336, 350, 391, 392, 393, 394, 397, 398, 399, 401, 406, 413, 419, 420, 421, 422, 423, 427, 428, 429, 430, 433, 435, 437, 440, 443, 444, 445, 452, 453, 455, 458, 460, 461, 465, 470, 471, 473, 474, 477, 492, 494, 496, 497, 506, 517, 519, 525, 528, 534, 547, 548, 567, ...], 'COP': [22257, 63897, 74959, 159878, 217161, 244427, 272170, 283473, 288349, 326013, 332702, 339601], 'PEN': [8032, 16638, 26780, 27224, 27389, 41199, 41200, 41203, 41205, 46052, 49193, 49194, 49197, 49199, 49202, 49203, 49990, 49991, 54414, 77161, 77162, 78909, 78910, 79585, 79588, 84729, 124063, 140118, 162992, 163029, 170714, 173680, 183694, 192881, 197224, 200024, 210597, 217507, 218716, 226097, 226099, 235804, 235806, 235807, 242778, 255306, 269383, 271126, 280612, 280613, 286988, 286989, 286990

NameError: name 'math' is not defined